File: 2443-check-if-there-is-a-valid-partition-for-the-array/2443-check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution {
public:
    int dp[1000001];
    
    bool solve(int i, vector<int>& nums) {
        if(i == nums.size()) {
            return dp[i] = true;
        }

        if(dp[i] != -1) return dp[i];

        // rule - 1
        if(i+1 < nums.size() && nums[i] == nums[i+1]){
            if(solve(i+2, nums)) {
                return dp[i] = true;
            }
        }

        // rule - 2
        if(i+2 < nums.size() && nums[i] == nums[i+1] && nums[i+1] == nums[i+2]) {
            if(solve(i+3, nums)) {
                return dp[i] = true;
            }
        }

        // rule - 3
        if(i+2 < nums.size() && nums[i+1] - nums[i] == 1 && nums[i+2] - nums[i+1] == 1) {
            if(solve(i+3, nums)) {
                return dp[i] = true;
            }
        }

        return dp[i] = false;
    }

    bool validPartition(vector<int>& nums) {
        memset(dp, -1, sizeof(dp));
        return solve(0, nums); 
    }  
};

File: 2443-check-if-there-is-a-valid-partition-for-the-array/2443-check-if-there-is-a-valid-partition-for-the-array.java
class Solution {
    int[] t = new int[100001];

    boolean solve(int i, int[] nums) {
        if(i == nums.length){
            t[i] = 1;
            return true;
        } 

        if(t[i] != -1) return t[i] == 1;

        if(i+1 < nums.length && nums[i] == nums[i+1]) {
            if(solve(i+2, nums)) {
                t[i] = 1;
                return true;
            }
        }

        if(i+2 < nums.length && nums[i] == nums[i+1] && nums[i+1] == nums[i+2]) {
            if(solve(i+3, nums)) {
                t[i] = 1;
                return true;
            }
        }

        if(i+2 < nums.length && nums[i+1] - nums[i] == 1 && nums[i+2] - nums[i+1] == 1) {
            if(solve(i+3, nums)) {
                t[i] = 1;
                return true;
            }
        }

        t[i] = 0;
        return false;
    }

    public boolean validPartition(int[] nums) {
        Arrays.fill(t, -1);
        return solve(0, nums);
    }
}